core-blocks/file/index.js

```diff
@@ -0,0 +1,172 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+import { createBlobURL } from '@wordpress/utils';
+import { createBlock } from '@wordpress/blocks';
+
+/**
+ * Internal dependencies
+ */
+import './style.scss';
+import edit from './edit';
+
+export const name = 'core/file';
+
+export const settings = {
+	title: __( 'File' ),
+
+	description: __( 'Link to a file.' ),
+
+	icon: 'media-default',
+
+	category: 'common',
+
+	keywords: [ __( 'document' ), __( 'pdf' ) ],
+
+	attributes: {
+		id: {
+			type: 'number',
+		},
+		href: {
+			type: 'string',
+		},
+		fileName: {
+			type: 'string',
+			source: 'text',
+			selector: 'a:not([download])',
+		},
+		// Differs to the href when the block is configured to link to the attachment page
+		textLinkHref: {
+			type: 'string',
+			source: 'attribute',
+			selector: 'a:not([download])',
+			attribute: 'href',
+		},
+		// e.g. `_blank` when the block is configured to open in a new window
+		textLinkTarget: {
+			type: 'string',
+			source: 'attribute',
+			selector: 'a:not([download])',
+			attribute: 'target',
+		},
+		showDownloadButton: {
+			type: 'boolean',
+			default: true,
+		},
+		downloadButtonText: {
+			type: 'string',
+			source: 'text',
+			selector: 'a[download]',
+			default: __( 'Download' ),
+		},
+	},
+
+	supports: {
+		align: true,
+	},
+
+	transforms: {
+		from: [
+			{
+				type: 'files',
+				isMatch: ( files ) => files.length === 1,
+				transform: ( files ) => {
+					const file = files[ 0 ];
+					const blobURL = createBlobURL( file );
+
+					// File will be uploaded in componentDidMount()
+					return createBlock( 'core/file', {
+						href: blobURL,
+						fileName: file.name,
+						textLinkHref: blobURL,
+					} );
```

Comment:
Here's what I got when dragging a small ZIP file into the editor:

<img width="700" alt="screen shot 2018-06-29 at 14 56 23" src="https://user-images.githubusercontent.com/150562/42097911-4e0b90fe-7bb1-11e8-8c79-c72b21d4ad5f.png">

Reply:
Fixed in 5202a06